hoodie-common/src/main/java/com/uber/hoodie/common/table/log/HoodieLogFormatReader.java

```diff
@@ -0,0 +1,201 @@
+  private final HoodieLogFile logFile;
+  private static final byte[] magicBuffer = new byte[4];
+  private final Schema readerSchema;
+  private HoodieLogBlock nextBlock = null;
+
+  HoodieLogFormatReader(FileSystem fs, HoodieLogFile logFile, Schema readerSchema, int bufferSize) throws IOException {
+    this.inputStream = fs.open(logFile.getPath(), bufferSize);
+    this.logFile = logFile;
+    this.readerSchema = readerSchema;
+  }
+
+  HoodieLogFormatReader(FileSystem fs, HoodieLogFile logFile, Schema readerSchema) throws IOException {
+    this(fs, logFile, readerSchema, DEFAULT_BUFFER_SIZE);
+  }
+
+  @Override
+  public HoodieLogFile getLogFile() {
+    return logFile;
+  }
+
+  private HoodieLogBlock readBlock() throws IOException {
+    // 2. Read the block type
+    int ordinal = inputStream.readInt();
+    Preconditions.checkArgument(ordinal < HoodieLogBlockType.values().length,
+        "Invalid block byte ordinal found " + ordinal);
+    HoodieLogBlockType blockType = HoodieLogBlockType.values()[ordinal];
+
+    // 3. Read the size of the block
+    int blocksize = inputStream.readInt();
+
+    // We may have had a crash which could have written this block partially
+    // Skip blocksize in the stream and we should either find a sync marker (start of the next block) or EOF
+    // If we did not find either of it, then this block is a corrupted block.
+    boolean isCorrupted = isBlockCorrupt(blocksize);
+    if(isCorrupted) {
+      return createCorruptBlock();
+    }
+
+    // 4. Read the content
+    // TODO - have a max block size and reuse this buffer in the ByteBuffer (hard to guess max block size for now)
+    byte[] content = new byte[blocksize];
+    inputStream.readFully(content, 0, blocksize);
+
+    switch (blockType) {
+      // based on type read the block
+      case AVRO_DATA_BLOCK:
+        return HoodieAvroDataBlock.fromBytes(content, readerSchema);
+      case DELETE_BLOCK:
+        return HoodieDeleteBlock.fromBytes(content);
+      case COMMAND_BLOCK:
+        return HoodieCommandBlock.fromBytes(content);
+      default:
+        throw new HoodieNotSupportedException("Unsupported Block " + blockType);
+    }
+  }
+
+  private HoodieLogBlock createCorruptBlock() throws IOException {
+    log.info("Log " + logFile + " has a corrupted block at " + inputStream.getPos());
+    long currentPos = inputStream.getPos();
+    long nextBlockOffset = scanForNextAvailableBlockOffset();
+    // Rewind to the initial start and read corrupted bytes till the nextBlockOffset
+    inputStream.seek(currentPos);
+    log.info("Next available block in " + logFile + " starts at " + nextBlockOffset);
+    int corruptedBlockSize = (int) (nextBlockOffset - currentPos);
+    byte[] content = new byte[corruptedBlockSize];
+    inputStream.readFully(content, 0, corruptedBlockSize);
+    return HoodieCorruptBlock.fromBytes(content);
+  }
+
+  private boolean isBlockCorrupt(int blocksize) throws IOException {
```

Comment:
instead of seeking once to find if its corrupt and seek again to read the data, can we determine the corrupt block from just the exception thrown when we attempt to read?